plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -143,26 +143,32 @@ object StmtConversionVisitor : FirVisitor<ExpEmbedding, StmtConversionContext>()
         val rightType = right.type
         // TODO: check whether isNullVal can be used here with no loss of generality.
         return if (leftType is NullableTypeEmbedding && rightType !is NullableTypeEmbedding) {
-            And(
-                NeCmp(left, leftType.nullVal),
-                // TODO: Replace the Eq comparison with a member call function to `left.equals`
-                EqCmp(left.withType(leftType.elementType), right.withType(leftType.elementType)),
-            )
-        } else if (leftType is NullableTypeEmbedding && rightType is NullableTypeEmbedding) {
-            Or(
-                And(
-                    EqCmp(left, leftType.nullVal),
-                    EqCmp(right, rightType.nullVal),
-                ),
-                // TODO: Replace the Eq comparison with a member call function to `left.equals`
+            share(left) { sharedLeft ->
                 And(
-                    And(
-                        NeCmp(left, leftType.nullVal),
-                        NeCmp(right, rightType.nullVal),
-                    ),
-                    EqCmp(left.withType(leftType.elementType), right.withType(leftType.elementType)),
-                ),
-            )
+                    NeCmp(sharedLeft, leftType.nullVal),
+                    // TODO: Replace the Eq comparison with a member call function to `left.equals`
+                    EqCmp(sharedLeft.withType(leftType.elementType), right.withType(leftType.elementType)),
+                )
+            }
+        } else if (leftType is NullableTypeEmbedding && rightType is NullableTypeEmbedding) {
```

Comment:
nit: I think that having a when expression here is better for readability:

```kotlin
return when {
    leftType is NullableTypeEmbedding && rightType !is NullableTypeEmbedding -> {
        // code ...
    }
    leftType is NullableTypeEmbedding && rightType is NullableTypeEmbedding -> {
        // code ...
    }
    else -> {
        // code ...
    }
}
```

WDYT?